backend/ipsec/handle_charon.go

```diff
@@ -0,0 +1,185 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipsec
+
+import (
+	"fmt"
+	"os/exec"
+	"time"
+
+	"github.com/coreos/flannel/Godeps/_workspace/src/github.com/bronze1man/goStrongswanVici"
+	log "github.com/coreos/flannel/Godeps/_workspace/src/github.com/golang/glog"
+	"github.com/coreos/flannel/subnet"
+)
+
+//TODO how to set debug level with standalone charon
+//TODO update depending on how we package
+const (
+	configFilePath = "/etc/strongswan.conf"
+	charonPath     = "/usr/local/libexec/ipsec/charon"
+	rekeyTime      = "10m"
+)
+
+type CharonIKEDaemon struct {
+	path           string
+	configFilePath string
+}
+
+func NewCharonIKEDaemon() (*CharonIKEDaemon, error) {
+	path, err := exec.LookPath(charonPath)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("Charon path: ", path)
+
+	return &CharonIKEDaemon{
+		path:           path,
+		configFilePath: configFilePath,
+	}, nil
+}
+
+func (charon *CharonIKEDaemon) RunCharonIKEDaemon() error {
+	cmd := exec.Command(charon.path)
+
+	return cmd.Run()
+}
+
+//TODO may not need to retry code in this function since loadcon is the first function to run.
+func (charon *CharonIKEDaemon) LoadSharedKey(remotePublicIP, password string) error {
+	var err error
+	var client *goStrongswanVici.ClientConn
+
+	for {
+		client, err = goStrongswanVici.NewClientConnFromDefaultSocket()
+		if err == nil {
+			break
+		} else {
+			log.Info("ClientConnection failed: ", err)
+			log.Infof("Retying in 1 second ...")
+			time.Sleep(1 * time.Second)
+		}
+	}
+	defer client.Close()
+
+	sharedKey := &goStrongswanVici.Key{
+		Typ:    "IKE",
+		Data:   password,
+		Owners: []string{remotePublicIP},
+	}
+
+	err = client.LoadShared(sharedKey)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Loaded shared key for: %v", remotePublicIP)
+	return nil
+}
+
+//TODO reqid as parameter
+func (charon *CharonIKEDaemon) LoadConnection(localLease, remoteLease *subnet.Lease, reqID, encap string) error {
+	var err error
+	var client *goStrongswanVici.ClientConn
+
+	for {
+		client, err = goStrongswanVici.NewClientConnFromDefaultSocket()
+		if err == nil {
+			break
+		} else {
+			log.Info("ClientConnection failed: ", err)
+			log.Infof("Retying in 1 second ...")
+			time.Sleep(1 * time.Second)
+		}
+	}
+	defer client.Close()
+
+	childConfMap := make(map[string]goStrongswanVici.ChildSAConf)
+	childSAConf := goStrongswanVici.ChildSAConf{
+		Local_ts:      []string{localLease.Subnet.String()},
+		Remote_ts:     []string{remoteLease.Subnet.String()},
+		ESPProposals:  []string{"aes256-sha256-modp4096"},
+		StartAction:   "start",
+		CloseAction:   "trap",
+		Mode:          "tunnel",
+		ReqID:         reqID,
+		RekeyTime:     rekeyTime,
```

Comment:
Good point. Instead of `margintime`, Vici defines `rand_time` which is the time range from which to choose a random value to subtract from rekey times. It defaults to `10%` of `rekey_time`. So in this case the rekey process could begin anytime after `9m` and before `10m` of the SA being established. We plan to change the `rekey_time` to the default of `1h` before merging. 

(the strongswan website isn't currently working so sharing the github page)
https://github.com/strongswan/strongswan/blob/9322e5b398efcf0a6f3bf576ef4b4b12b5ae6528/src/swanctl/swanctl.opt